### PR TITLE
NestedFolders: Nested folder picker

### PR DESCRIPTION
**Pull request by joshhunt**
Picking up from @tskarhed's work in https://github.com/grafana/grafana/pull/68124

Adds a new NestedFolderPicker component to pick folders in a nested structure. The intention of this component is that it will completely replace the existing folder picker components (which we can do 'easily' without the need to worry about making breaking changes), but because of potential difference in API, and that we'll iterate on this component over multiple PRs, we've decided to feature flag this specifically and gradually replace instances. We intend for this feature flag to be very 'short lived' once the component is actually usable :)

image

Still pretty early:
 - Lists folders
 - Clicking expand button expands that folder
 - Empty folders just show.... nothing.
 - Clicking a row that _isnt_ in chevron selects that folder
 - Is an 'uncontrolled' component - it's up to the consumer to maintain selection state
 
Not implemented yet:
 - Search
 - Pagination (probably seperate PR)

It's currently being used in:
 - Move folder dialog
 - Save new dashboard
 - Dashboard Settings

### CODE REVIEW

File: public/app/features/browse-dashboards/state/hooks.ts [hunk 1]
```diff
@@ -126,19 +126,27 @@ export function useLoadNextChildrenPage(folderUID: string | undefined) {
...
  */
-function createFlatTree(
+export function createFlatTree(
   folderUID: string | undefined,
   rootCollection: BrowseDashboardsState['rootItems'],
   childrenByUID: BrowseDashboardsState['childrenByParentUID'],
   openFolders: Record<string, boolean>,
-  level = 0
+  level = 0,
+  insertEmptyFolderIndicator = true
```
Comment (ashharrison90) [inline public/app/features/browse-dashboards/state/hooks.ts:135]: is this needed anymore?

File: public/app/features/browse-dashboards/types.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { CellProps, Column, HeaderProps } from 'react-table';
 import { DashboardViewItem, DashboardViewItemKind } from 'app/features/search/types';
 
 export type DashboardTreeSelection = Record<DashboardViewItemKind, Record<string, boolean | undefined>> & {
-  $all: boolean;
+  $all?: boolean;
```
Comment (ashharrison90) [inline public/app/features/browse-dashboards/types.ts:6]: why is this now optional? 🤔

**Review by ashharrison90 — COMMENTED** [review]
looks like a great start! 🥳 

noticed a couple of things:
- in the move modal and save as drawer, selecting a folder doesn't show the indicator for which folder is selected (it does in general settings)
- there's no way to select the root folder

added some minor nits/questions

File: public/app/features/browse-dashboards/state/hooks.ts (hunk 1 above)
Comment (joshhunt) [reply public/app/features/browse-dashboards/state/hooks.ts:135]: Folder picker uses it to not get the "empty folder" rows https://github.com/grafana/grafana/blob/joshhunt/nested-folder-picker/public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx#L62

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx [hunk 3]
```diff
@@ -28,10 +31,11 @@ export function GeneralSettingsUnconnected({
...
   const [renderCounter, setRenderCounter] = useState(0);
 
-  const onFolderChange = (folder: { uid: string; title: string }) => {
-    dashboard.meta.folderUid = folder.uid;
-    dashboard.meta.folderTitle = folder.title;
+  const onFolderChange = (newFolder: FolderChange) => {
+    dashboard.meta.folderUid = newFolder.uid === ROOT_FOLDER ? '' : newFolder.uid;
+    dashboard.meta.folderTitle = newFolder.title;
     dashboard.meta.hasUnsavedFolderChange = true;
+    setRenderCounter(renderCounter + 1);
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:38]: Yeah it's gross, but it's what all the other properties that mutate dashboard do so it seems to fit?

File: public/app/features/browse-dashboards/types.ts (hunk 2 above)
Comment (joshhunt) [reply public/app/features/browse-dashboards/types.ts:6]: thanks for catching this. i... can't remember, i can't think of a reason why. i undid the change and everything still seems fine so 🤷🏻

File: public/app/features/browse-dashboards/state/hooks.ts (hunk 1 above)
Comment (ashharrison90) [reply public/app/features/browse-dashboards/state/hooks.ts:135]: ahhhhh got it, that was the logic i was missing ty 👍

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:38]: yeah i understand, it's just using the already existing icky pattern 😄

**Comment by joshhunt** [comment]
@ashharrison90 if you get the chance, would you mind checking this out again. Apart from the smaller cleanups, changes in here:

 - Root/General folder now selectable.
 - Selection works in MoveFolder
   - One big difference with this new component vs old folder picker is that this is a fully uncontrolled component - it's up to the parent to pass in `value` properly when the selection changes
 - Now only the folder name is clickable to select it to prevent nested click targets

**Review by ashharrison90 — APPROVED** [review]
bravo 👏 

couple of thoughts, can tackle (or not) in follow up PRs:
- we probably should truncate in this scenario:
image
- i find the `Dashboards` item being at the same level as the other folders pretty confusing tbh. my gut is to have it as the parent for the folders and expanded by default. i realise that immediately introduces a level of nesting which is annoying but it's clear where it is in the hierarchy at least.

**Pull request merged**